### ISSUE TITLE
nn.InstanceNorm and nn.GroupNorm are affected by padding, so they need to masking

### ISSUE THREAD

**Issue by kss2517** [pytorch/pytorch#81985]
### 🐛 Describe the bug

For batch_size > 1, variable-length inputs (e.g. speech, text) are padded in order to construct one batch tensor.

**When this tensor goes through nn.InstanceNorm series (nn.InstanceNorm1d, nn.InstanceNorm2d, ...) or nn.GroupNorm,
the calculations of mean and variance take the padded dummies altogether.**

This causes harmful batch inconsistency problem.

**So, nn.InstanceNorm and nn.GroupNorm must be modified to take valid input lengths and to mask out the padded dummies during calculating mean and variance.**

Sorry for not attaching example code.

### Versions

PyTorch version: 1.12.0

**Comment by vadimkantorov** [comment]
Also support this feature request. Maybe it can be supported by letting *Norm modules support NestedTensors? @cpuhrsch 

some of my impls:
```python
class MaskedInstanceNorm1d(nn.InstanceNorm1d):
	def __init__(self, *args, temporal_mask = False, legacy = True, **kwargs):
		super().__init__(*args, **kwargs)
		self.temporal_mask = temporal_mask
		self.legacy = legacy

	def forward(self, x, mask = None):
		if not self.temporal_mask or mask is None:
			if self.legacy:
				assert self.track_running_stats is False
				# BUG: https://github.com/pytorch/pytorch/issues/44636
				#std, mean = torch.std_mean(x, dim = -1, keepdim = True)
				# replaced with new version for .onnx export fix!

				# BUG failed:Type Error: Type parameter (T) bound to different types (tensor(float16) and tensor(float) in node (Mul_88). Mul_88 node located in torch.std function.
				#std = torch.std(x, dim=-1, keepdim=True)
				mean = torch.mean(x, dim=-1, keepdim=True)
				# NOTE: Also took a CER WER measurement on a validation set and did not notice the difference between torch.std(x, dim=-1, keepdim=True) and bottom std setup.
				# torch.std uses Bessel unbiased estimate by default, code below performs biased estimation.
				x_minus_mean = x - mean
				# NOTE: pow(2) cause onnxruntime warnings like "CUDA kernel not found in registries for Op type: Pow node name: Pow_76" To avoid this pow replaced by x*x
				std = ((x_minus_mean * x_minus_mean).mean(dim=-1, keepdim=True)).sqrt()
				return x_minus_mean / (std + self.eps)
			else:
				return super().forward(x)
		else:
			assert self.track_running_stats is False
			xlen = mask.int().sum(dim = -1, keepdim = True)
			mean = (x * mask).sum(dim = -1, keepdim = True) / xlen
			zero_mean_masked = mask * (x - mean)
			std = ((zero_mean_masked * zero_mean_masked).sum(dim = -1, keepdim = True) / xlen).sqrt()
			return zero_mean_masked / (std + self.eps)
```

**Comment by cpuhrsch** [comment]
cc @jbschlosser

**Comment by jbschlosser** [comment]
We're broadening op support for nested tensors as we speak, and LayerNorm at least is high-pri due to its usage within Transformers.

**Comment by vadimkantorov** [comment]
@cpuhrsch can one say that for *C logical layout, nested tensor physical representation after flatten+cat will also be *C? Some ops (such as LayerNorm) can then be applied directly to this physical representation without any unpacking/repacking

**Comment by albanD** [comment]
@ezyang why the pt2 label?

**Comment by albanD** [comment]
High pri part is to add NestedTensor support to these functions.
It is still an open question if we want to do anything beyond that.

**Comment by vadimkantorov** [comment]
Another problem of support via nestedtensor could be that it might be a lot of work for having it exported correctly to ONNX (and maybe some other backends)

**Comment by youkaichao** [comment]
It is strange that this problem is not solved in 2022. Consider an input in NLP with shape ``[N, L, E]``, ``L`` is actually the maximum sequence length, and there may be many paddings in the tensor. Currently, the layernorm implementation would be unfriendly for short sequences in the batch.

It seems [nested tensors](https://pytorch.org/docs/stable/nested.html) are only supported in inference mode, and making layernorm support nested tensors does not solve the problem during training.

**Comment by vadimkantorov** [comment]
Well, for LayerNorm there's less of this problem because it typically normalizes across the channel dimension, so every token is processed independently and pad tokens do not affect others. For affine transform though, the pad tokens affect the learned parameters and the gradients...

**Comment by youkaichao** [comment]
> Well, for LayerNorm there's less of this problem because it typically normalizes across the channel dimension, so every token is processed independently and pad tokens do not affect others. For affine transform built into LayerNorm though, the pad tokens affect the learned parameters and the gradients...

Oh, my bad, I am often confused by the image in [https://pytorch.org/docs/stable/generated/torch.nn.LayerNorm.html](https://pytorch.org/docs/stable/generated/torch.nn.LayerNorm.html) and think LN applies to the [L, E] dimensions...
![](https://pytorch.org/docs/stable/_images/layer_norm.jpg)

I think you are right, the pad tokens will affect gradients of weight and bias in LN. If $y=\Sigma_{i=1}^{N} w x_i + b$ and there are pad tokens in it, we should apply the mask to both w and b: $y=\Sigma_{i=1}^{N} (w x_i + b)*m_i$. From this perspective, it seems we only need to apply a mask to the output so that the gradients are corrected :)

**Comment by youkaichao** [comment]
> Well, for LayerNorm there's less of this problem because it typically normalizes across the channel dimension, so every token is processed independently and pad tokens do not affect others. For affine transform built into LayerNorm though, the pad tokens affect the learned parameters and the gradients...

@vadimkantorov  It seems both forward and backward of LayerNorm are okay, as long as those padding tokens are not involved in the final loss calculation. I made a tiny working example in the following:

```python
import torch

from torch import nn

# a block with self-attention and layernorm
class EncoderBlock(nn.Module):
    def __init__(self):
        super().__init__()
        self.self_attention = nn.MultiheadAttention(64, 8, dropout=0, batch_first=True)
        self.ln_1 = nn.LayerNorm(64, eps=1e-6)
        self.mlp = nn.Linear(64, 10)

    def forward(self, x, key_padding_mask):
        x, _ = self.self_attention(query=x, key=x, value=x, need_weights=False, key_padding_mask=key_padding_mask)
        x = self.ln_1(x)
        x = self.mlp(x)
        return x

block = EncoderBlock()

params = list(block.parameters())

random_input = torch.randn(64, 128, 64, dtype=torch.float32)

random_target = torch.randn(64, 128, 10, dtype=torch.float32)

random_mask = random_input[:, :, 0] > 0 # shape [64, 128]

output = block(random_input, random_mask)

loss1 = (((random_target - output) * random_mask.view(64, 128, 1).logical_not()) ** 2).sum()
loss1.backward()

grad1 = [p.grad.clone() for p in params]

# perturb the pad tokens
perturbed_input = torch.randn(64, 128, 64, dtype=torch.float32) * random_mask.view(64, 128, 1) + random_input
output2 = block(perturbed_input, random_mask)

loss2 = (((random_target - output2) * random_mask.view(64, 128, 1).logical_not()) ** 2).sum()
block.zero_grad()
loss2.backward()

grad2 = [p.grad.clone() for p in params]

print('output differs')
print(((output - output2)**2).sum().item()) # 2786.26123046875 by my laptop

diff = (((output - output2) * random_mask.view(64, 128, 1).logical_not())**2).sum().item()
print('but locations without masking are the same')
print(diff) # 0 by my laptop

grad_diff = sum([((a - b) ** 2).sum().item() for a, b in zip(grad1, grad2)], 0)
print('and gradients are the same, so long as we do not use pad tokens for calculating the loss')
print(grad_diff) # 0 by my laptop
```

**Comment by vadimkantorov** [comment]
> as long as those padding tokens are not involved in the final loss calculation

If there is some "token mixing" of some sort or some layers having global receptive field (like unmasked attention), this wouldn't help I guess...

**Comment by vadimkantorov** [comment]
> I think it is the user's responsibility to be aware of the mask when they use global receptive field. Maybe we can mention it somewhere in the documentation.

For LayerNorm - agreed

But if more modules like InstanceNorm and GroupNorm will get masked support (via nestedtensor or a more explicit api), I think it would make sense to support it also for LayerNorm

**Comment by ydwu4** [comment]
Hi, is this issue still relevant?

**Comment by yf225** [comment]
This seems not torch.compile related, removing `oncall: pt2` label.

**Comment by ysurs** [comment]
Hey, I faced a problem using nn.InstanceNorm1d while a batched input for TTS application. As rightly pointed here https://github.com/pytorch/pytorch/issues/81985#issue-1315103272 .

I have currently written a custom function with masking but it seems that the performance has dropped, any plans for a fix or can you recommend any work around to make batching work ?

Hey @vadimkantorov , this is still relevant right, there is no other way but to implement custom masked instancenorm ?
